lib/cog/pipeline.ex

```diff
@@ -0,0 +1,337 @@
+      request = Keyword.fetch!(opts, :request) |> sanitize_request
+      owner = if policy in [:owner, :adapter_owner] do
+        Keyword.fetch!(opts, :owner)
+      else
+        nil
+      end
+      case ConnectionSup.connect() do
+        {:ok, conn} ->
+          {:ok, %__MODULE__{policy: policy, owner: owner,
+                            request: request, conn: conn,
+                            token: Tokens.new(), status: :running}}
+        error ->
+          Logger.error("Failed to connect pipeline #{request.id} to message bus: #{inspect error}")
+          {:stop, error}
+      end
+    rescue
+      e in KeyError ->
+        {:stop, {:missing_option, e.key}}
+    end
+  end
+
+  def handle_call(:run, _from, state) do
+    state = %{state | started: DateTime.utc_now()}
+    case fetch_user(state) do
+      {:ok, user} ->
+        with {:ok, parsed, destinations} <- parse(user, state),
+             {:ok, perms} <- PermissionsCache.fetch(user),
+               user_json <- EctoJson.render(user) do
+          start_pipeline(parsed, destinations, user_json, perms, state)
+        else
+          {:error, {:parse_error, message}} -> start_error_pipeline({:error, :parse_error, message}, user, state)
+          error -> start_error_pipeline(error, user, state)
+        end
+      {:error, :user_not_found}=error ->
+        start_error_pipeline("unregistered-user", error, nil, state)
+      error ->
+        start_error_pipeline(error, nil, state)
+    end
+  end
+  def handle_call(_msg, _from, state) do
+    {:reply, :ignored, state}
+  end
+
+  def handle_cast(:teardown, state) do
+    duration = DateTime.to_unix(DateTime.utc_now, :milliseconds) - DateTime.to_unix(state.started, :milliseconds)
+    Logger.info("Pipeline #{state.request.id} ran for #{duration} ms")
+    Enum.each(state.stages, &(Process.send(&1, {:pipeline_complete, self()}, [])))
+    {:noreply, %{state | status: :done}}
+  end
+  def handle_cast(_, state) do
+    {:noreply, state}
+  end
+
+  # Close bus connection and exit after last stage exits
+  # This avoids closing the connection while one of the sinks is still using it
+  def handle_info({:DOWN, _, _, stage, _}, %__MODULE__{status: :done, stages: [stage]}=state) do
+    Logger.debug("Pipeline #{state.request.id} terminated")
+    Connection.disconnect(state.conn)
+    {:stop, :normal, %{state | stages: []}}
+  end
+  def handle_info({:DOWN, _, _, stage, _}, %__MODULE__{status: :done, stages: stages}=state) do
+    {:noreply, %{state | stages: List.delete(stages, stage)}}
+  end
+  # Change pipeline state to :done if a stage crashes
+  def handle_info({:DOWN, _, _, stage, reason}, %__MODULE__{status: :running}=state) do
+    if stage in state.stages do
+      message = case reason do
+                  {_, stacktrace} ->
+                    "Pipeline #{state.request.id} crashed: #{inspect stacktrace, pretty: true}"
+                  _ ->
+                    "Pipeline #{state.request.id} crashed"
```

Comment:
That's the problem. It's huge. At least 2 screens worth of text. Could be handled by adding `String.Chars` impls for a few structs. I could tackle that in a separate PR.